boot/seal.go

```diff
@@ -25,61 +25,327 @@ import (
+	}
+
+	pbc := toPredictableBootChains(append(runModeBootChains, recoveryBootChain))
+
+	// XXX: store the predictable bootchains
+
+	// get parameters from bootchains and seal the key
+	params, err := sealKeyParams(pbc)
+	if err != nil {
+		return fmt.Errorf("cannot build key sealing parameters: %v", err)
+	}
+
+	if err := secbootSealKey(key, params); err != nil {
+		return fmt.Errorf("cannot seal the encryption key: %v", err)
+	}
+
+	return nil
+}
+
+func buildRecoveryBootChain(rbl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv) (bc bootChain, err error) {
+	// get the command line
+	cmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	if err != nil {
+		return bc, fmt.Errorf("cannot obtain recovery kernel command line: %v", err)
+	}
+
+	// get kernel information from seed
+	perf := timings.New(nil)
+	_, snaps, err := seedReadSystemEssential(dirs.SnapSeedDir, modeenv.RecoverySystem, []snap.Type{snap.TypeKernel}, perf)
+	if err != nil {
+		return bc, err
+	}
+	if len(snaps) != 1 {
+		return bc, fmt.Errorf("cannot obtain recovery kernel snap")
+	}
+	seedKernel := snaps[0]
+
+	var kernelRev string
+	if seedKernel.SideInfo.Revision.Store() {
+		kernelRev = seedKernel.SideInfo.Revision.String()
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildRecoveryAssetChain(rbl, modeenv)
+	if err != nil {
+		return bc, err
+	}
+
+	return bootChain{
+		BrandID:        model.BrandID(),
+		Model:          model.Model(),
+		Grade:          string(model.Grade()),
+		ModelSignKeyID: model.SignKeyID(),
+		AssetChain:     assetChain,
+		Kernel:         seedKernel.Path,
+		KernelRevision: kernelRev,
+		KernelCmdline:  cmdline,
+		model:          model,
+		blName:         rbl.Name(),
+		kernelBootFile: bootloader.NewBootFile(seedKernel.Path, kbf.Path, kbf.Role),
+	}, nil
+}
+
+func buildRunModeBootChains(rbl, bl bootloader.Bootloader, model *asserts.Model, modeenv *Modeenv) ([]bootChain, error) {
+	// get the command line
 	cmdline, err := ComposeCandidateCommandLine(model)
 	if err != nil {
-		return fmt.Errorf("cannot obtain kernel command line: %v", err)
+		return nil, fmt.Errorf("cannot obtain kernel command line: %v", err)
+	}
+
+	// get asset chains
+	assetChain, kbf, err := buildRunModeAssetChain(rbl, bl, modeenv)
+	if err != nil {
+		return nil, err
+	}
+
+	// get run mode kernels
+	runModeKernels, err := runModeKernelsFromModeenv(modeenv)
+	if err != nil {
+		return nil, err
 	}
 
-	// Get the expected kernel command line of the recovery system we're installing from
-	recoveryCmdline, err := ComposeRecoveryCommandLine(model, modeenv.RecoverySystem)
+	chains := make([]bootChain, 0, 2)
+	for _, k := range runModeKernels {
+		chains = append(chains, bootChain{
+			BrandID:        model.BrandID(),
+			Model:          model.Model(),
+			Grade:          string(model.Grade()),
+			ModelSignKeyID: model.SignKeyID(),
+			AssetChain:     assetChain,
+			Kernel:         k,
+			// XXX: obtain revision
```

Comment:
for kernels out of modeenv  current_kernels we can use snap.ParsePlaceInfoFromSnapFileName , it's used already in a couple of place in boot